### PR TITLE
core/state: avoid update slotsToPrefetch when prefetch is false

### PR DESCRIPTION
**Pull request by zhiqiangxu**

### CODE REVIEW

File: core/state/state_object.go [hunk 1]
```diff
@@ -310,16 +310,23 @@ func (s *stateObject) setState(key, value common.Hash) {
...
+			if value != s.originStorage[key] {
+				slotsToPrefetch = append(slotsToPrefetch, common.CopyBytes(key[:])) // Copy needed for closure
+			}
+		}
+		if len(slotsToPrefetch) > 0 {
+			s.db.prefetcher.prefetch(s.data.Root, slotsToPrefetch)
+		}
+	} else {
+		for key, value := range s.dirtyStorage {
+			s.pendingStorage[key] = value
```
Comment (s1na) [inline core/state/state_object.go:326]: I slightly prefer moving this to the beginning of the func instead of adding an `else` clause

**Review by s1na — COMMENTED** [review]
Looks logically equivalent

File: core/state/state_object.go (hunk 1 above)
Comment (zhiqiangxu) [reply core/state/state_object.go]: Updated

**Review by s1na — APPROVED** [review]
LGTM

File: core/state/state_object.go (hunk 1 above)
Comment (zhiqiangxu) [reply core/state/state_object.go:326]: @s1na Weird, the above change causes build failure, I'm reverting it to see whether it will pass

Comment (s1na) [reply core/state/state_object.go:326]: Restarting the build for that commit to make sure it wasn't because of a flaky test

Comment (s1na) [reply core/state/state_object.go:326]: so it passed the second time around

Comment (zhiqiangxu) [reply core/state/state_object.go:326]: Cool, let me recover the code!

Comment (s1na) [reply core/state/state_object.go]: lolling at the commit title `Revert "Revert "fix for review""` :)) sorry we went so much back and forth for a nitpick

**Review by holiman — COMMENTED** [review]
I'm sceptical. This makes the case where prefetching is disabled slightly faster, but makes the case where prefetching is enabled slightly slower. 

In almost all circumstances, prefetching is enabled. This prefetcher, the trie prefetcher, loads the tries after the tx is executed, so that when we eventually reach the end of block and want to do the actual trie _writes_, the paths are already resolved. 

I don't see why we would want to optimize the abnormal-path over the normal-path?

**Comment by zhiqiangxu** [comment]
> I'm sceptical. This makes the case where prefetching is disabled slightly faster, but makes the case where prefetching is enabled slightly slower.
> 
> In almost all circumstances, prefetching is enabled. This prefetcher, the trie prefetcher, loads the tries after the tx is executed, so that when we eventually reach the end of block and want to do the actual trie _writes_, the paths are already resolved.
> 
> I don't see why we would want to optimize the abnormal-path over the normal-path?

I've reverted to the first version so that there's no performance loss.

**Review by holiman — APPROVED** [review]
I guess this is ok. I'm not sure whether it'll make any difference, AFAIC we always prefetch on the blocks that are imported, this might shave a bit of allocations from the case where we execute things like the pending state or when we execute due to rpc requests. 

I'm not fully certain that the increased LOC is worth it, but my "Approve" means that I think it's functionally correct.

**Pull request closed without merging**